ql/src/java/org/apache/hadoop/hive/ql/ddl/table/misc/truncate/TruncateTableAnalyzer.java

```diff
@@ -338,4 +370,56 @@ private void addStatTask(ASTNode root, Table table, Path oldPartitionLocation, P
       moveTask.addDependentTask(statTask);
     }
   }
+
+  public StringBuilder constructDeleteQuery(Table table, Map<String, String> partitionSpec) throws SemanticException {
+    String qualifiedTableName = HiveTableName.ofNullable(HiveUtils.unparseIdentifier(table.getTableName(), this.conf),
+            HiveUtils.unparseIdentifier(table.getDbName(), this.conf), null).getNotEmptyDbTable();
+    StringBuilder sb = new StringBuilder().append("delete from ").append(qualifiedTableName)
+            .append(" where ");
+    Deserializer deserializer = table.getDeserializer();
+    Map<String, PrimitiveObjectInspector.PrimitiveCategory> stringTypeInfoMap = new HashMap<>();
+    try {
+      ObjectInspector objectInspector = deserializer.getObjectInspector();
+      if (objectInspector.getCategory() == ObjectInspector.Category.STRUCT) {
+        StructObjectInspector structObjectInspector = (StructObjectInspector) objectInspector;
+        List<? extends StructField> structFields =  structObjectInspector.getAllStructFieldRefs();
+        for (StructField structField : structFields) {
+          if (structField.getFieldObjectInspector().getCategory() == ObjectInspector.Category.PRIMITIVE) {
+            PrimitiveObjectInspector primitiveObjectInspector = (PrimitiveObjectInspector) structField.getFieldObjectInspector();
+            stringTypeInfoMap.put(structField.getFieldName(),
+                primitiveObjectInspector.getTypeInfo().getPrimitiveCategory());
+          }
+        }
+      }
+    } catch (SerDeException e) {
+      throw new SemanticException(String.format("Unable to get object inspector due to: %s", e));
+    }
+    boolean first = true;
+    for (String key : partitionSpec.keySet()) {
+      PrimitiveObjectInspector.PrimitiveCategory category = stringTypeInfoMap.get(key);
+      if (category == null) {
+        throw new SemanticException(String.format(
+            "Only primitive partition column type is supported via TRUNCATE operation " +
+            " but column %s is not a primitive category.", key));
+      }
+      String value = partitionSpec.get(key);
+      boolean shouldEncloseQuotes = TypeInfoUtils.shouldEncloseQuotes(category);
+      if (value != null) {
+        sb.append(first ? "" : " and ").append(key).append(" = ");
+        if (shouldEncloseQuotes) {
+          // The partition value is containing a single-quote hence we cannot use \' as the enclosing quote.
+          // We need to use \" as the enclosing quote.
+          if (value.contains("'")) {
+            sb.append("\"" ).append(value).append("\"");
+          } else {
+            sb.append("'").append(value).append("'");
+          }
+        } else {
+          sb.append(value);
+        }
```

Comment:
Sorry I thought that we are escaping identifiers here but that is not the case. These are actually string literals. I don't know if there is any utility method to escape them.

The escape character for `'` is `\'` and for `"` is `\"`. I prefer `'` for enclosing because that is the standard.

Using both characters in literals are also accepted:
```
create table t1(a string, b int);
insert into t1(a,b) values ('foo\'\'bar', 1), ("foo\"\"bar", 2), ("foo'\"bar", 3), ('foo\'"bar', 4);
```
How about 
```
sb.append("'").append(value.replace("'", "\\'")).append("'");
```
to cover that case when both are present?

Reply:
Added the same. Done.